keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,154 @@
+
+### Convert managedFields directly to internal fieldpath
+
+On updates and applies, the apiserver needs to convert managedFields into an internal representation. We originally decoupled the managedFields API from the internal representation of a fieldset, to allow arbitrary changes to the internal representation without breaking the API. However converting from a serialized config to a gostruct and then into a different struct causes a lot of allocations of small objects and stress on the garbage collector. Instead, we can directly serialize as the internal representation.
+
+The internal format could be stored in protobuf, which would make it much smaller than json (json has a lot of repeated fields that make it extremely large).
+
+### Only update managedFields if previously applied to
+
+Most objects created with post or put are never applied to, so tracking managedFields on them is a waste of resources. If an object without managedFields does get applied to, we can still generate a single managedFields entry for all the existing fields. The generated managedFields conflicts set will be the same, but we won’t have any way of knowing know who is being conflicted with. CPU and memory cost on the apiserver will be dependent on the number of objects managed by server-side apply. It would increase more gradually as controllers decided to adopt SSA. Adoption could be decided on a controller by controller basis, weighing the cost vs. benefits each time. If someone applies to an object created by post or put, the conflicts they see would not contain information about the timestamp or field manager, so conflicts become less useful.
+
+### Storing in one format, serving in another
+
+Another possible solution would be to combine many of these solutions:
+
+Some combination of storing the field managers as one big blob in etcd, using the internal fieldpath representation, as gzipped protobuf, would give us a much lower overhead in etcd and very fast serialization/deserialization. This is especially true if deserialization doesn’t need to know about the specifics of managedfields (hence just deserializing as a blob).
+
+Serving the API as it currently is (or maybe changing it if we believe we need to change what we serve), but only on demand so that the price of deserializing would only be paid by clients that specifically request it.
+
+Based on initial numbers, this solution has the most advantages:
+etcd objects (based on pods above) only grow by 30%
+listing objects has nearly zero cost since we don’t send unless requested,
+patches and updates can deserialize the data very quickly directly in protobuf, and process the data
```

Comment:
If you did this as proto, how does it help CustomResources?

Reply:
Storing it as a blob, we can still use the proto form of managedfields, even if the object containing the blob is in json/yaml or whatever format, if we published the proto for the type.